### PR TITLE
[hotfix] Extend Optimizer + update doc

### LINKED ISSUE #5092: Optimization docs and override optimizer_step docs
## 📚 Documentation

Optimization docs need to be updated with the correct way of overriding ```optimizer_step()``` method, which now takes in a closure as a non-optional parameter.

https://pytorch-lightning.readthedocs.io/en/latest/optimizers.html#step-optimizers-at-arbitrary-intervals

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #5092

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -491,16 +491,14 @@ def optimizer_step(self, optimizer, opt_idx, batch_idx, train_step_and_backward_
 
         # model hook
         model_ref.optimizer_step(
-            epoch=self.trainer.current_epoch,
-            batch_idx=batch_idx,
-            optimizer=optimizer,
-            optimizer_idx=opt_idx,
-            optimizer_closure=train_step_and_backward_closure,
+            self.trainer.current_epoch,
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:494]: why do this? If it's explicit it reduces the risk of potential errors in ordering imo

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1254,7 +1252,7 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
         if not isinstance(optimizer, LightningOptimizer):
             # wraps into LightingOptimizer only for running step
             optimizer = LightningOptimizer.to_lightning_optimizer(optimizer, self.trainer)
-        optimizer.step(closure=optimizer_closure, *args, **kwargs)
+        optimizer.step(closure=optimizer_closure)
```
Comment (SeanNaren) [inline pytorch_lightning/core/lightning.py:1255]: Could you explain how the user can provide args/kwargs to his optimizer now? I assume directly to the lightning optimizer?

File: pytorch_lightning/__init__.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 """Root package info."""
 
-__version__ = '1.1.0'
+__version__ = "20201211"
```
Comment (tchaton) [inline pytorch_lightning/__init__.py:3]:
```suggestion
__version__ = '1.1.0'
```

File: setup.py [hunk 4]
```diff
@@ -61,7 +61,7 @@
 # the goal of the project is simplicity for researchers, don't want to add too much
 # engineer specific practices
 setup(
-    name="pytorch-lightning",
+    name="pytorch-lightning-nightly",
```
Comment (tchaton) [inline setup.py:64]:
```suggestion
    name="pytorch-lightning",
```

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:494]: Because It was breaking backward compatibility.

Comment (SeanNaren) [reply pytorch_lightning/trainer/training_loop.py:494]: thanks!

File: docs/source/optimizers.rst [hunk 5]
```diff
@@ -191,46 +191,68 @@ override the :meth:`optimizer_step` function.
 
 For example, here step optimizer A every 2 batches and optimizer B every 4 batches
 
-.. testcode::
+.. note:: When using Trainer(enable_pl_optimizer=True), no need to call `.zero_grad`.
```
Comment (Borda) [inline docs/source/optimizers.rst:194]:
```suggestion
.. note:: When using ``Trainer(enable_pl_optimizer=True)``, no need to call ``.zero_grad(...)``.
```

File: docs/source/optimizers.rst [hunk 6]
```diff
@@ -191,46 +191,68 @@ override the :meth:`optimizer_step` function.
...
+        if optimizer_i == 1:
+            optimizer.step(closure=closure, make_optimizer_step=batch_nb % 4 == 0)
 
 Here we add a learning-rate warm up
 
 .. testcode::
 
     # learning rate warm-up
-    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, second_order_closure=None, on_tpu=False, using_native_amp=False, using_lbfgs=False):
+    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
```
Comment (Borda) [inline docs/source/optimizers.rst:234]: not sure if it make sense, but shall we include typing also in docs examples?

File: docs/source/optimizers.rst [hunk 7]
```diff
@@ -191,46 +191,68 @@ override the :meth:`optimizer_step` function.
...
+               optimizer.step(closure=closure)
 
         # update discriminator opt every 4 steps
         if optimizer_i == 1:
             if batch_nb % 4 == 0 :
-                optimizer.step()
-                optimizer.zero_grad()
+               optimizer.step(closure=closure)
+
+
```
Comment (Borda) [inline docs/source/optimizers.rst:213]: let's add a sentence what is this example about, how it differs from the one above?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/core/lightning.py:1255]: cc: @awaelchli @justusschock

**Review by ananyahjha93 — APPROVED** [review]

File: docs/source/optimizers.rst (hunk 5 above)
Comment (awaelchli) [inline docs/source/optimizers.rst:194]:
```suggestion
.. note:: When using Trainer(enable_pl_optimizer=True), there is no need to call `.zero_grad()`.
```

File: docs/source/optimizers.rst (hunk 6 above)
Comment (tchaton) [reply docs/source/optimizers.rst:236]: Not sure, as we would have to import Optimizer, callable, etc..

File: docs/source/optimizers.rst (hunk 7 above)
Comment (tchaton) [reply docs/source/optimizers.rst:215]: Done

File: docs/source/optimizers.rst [hunk 8]
```diff
@@ -191,46 +191,70 @@ override the :meth:`optimizer_step` function.
...
 
         # update discriminator opt every 4 steps
         if optimizer_i == 1:
             if batch_nb % 4 == 0 :
-                optimizer.step()
-                optimizer.zero_grad()
+               optimizer.step(closure=closure)
+
+.. note:: When using ``Trainer(enable_pl_optimizer=True)``, ``.step`` accept a boolean ``make_optimizer_step`` which 
+can be used as below.
```
Comment (awaelchli) [inline docs/source/optimizers.rst]:
```suggestion
.. note:: When using ``Trainer(enable_pl_optimizer=True)``, ``.step`` accepts a boolean ``make_optimizer_step`` which 
can be used as shown below.
```

File: docs/source/optimizers.rst [hunk 9]
```diff
@@ -191,46 +191,70 @@ override the :meth:`optimizer_step` function.
...
+.. testcode::
+
+    def optimizer_zero_grad(self, current_epoch, batch_idx, optimizer, opt_idx):
+      optimizer.zero_grad()
+
+    # Alternating schedule for optimizer steps (ie: GANs)
+    def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, closure, on_tpu=False, using_native_amp=False, using_lbfgs=False):
+        # update generator opt every 2 steps
+        if optimizer_i == 0:
+            optimizer.step(closure=closure, make_optimizer_step=batch_nb % 2 == 0)
```
Comment (awaelchli) [inline docs/source/optimizers.rst:224]:
```suggestion
            optimizer.step(closure=closure, make_optimizer_step=(batch_nb % 2) == 0)
```

File: docs/source/optimizers.rst [hunk 10]
```diff
@@ -191,46 +191,70 @@ override the :meth:`optimizer_step` function.
...
 
-        # ...
-        # add as many optimizers as you want
+        # update discriminator opt every 4 steps
+        if optimizer_i == 1:
+            optimizer.step(closure=closure, make_optimizer_step=batch_nb % 4 == 0)
```
Comment (awaelchli) [inline docs/source/optimizers.rst:228]:
```suggestion
            optimizer.step(closure=closure, make_optimizer_step=(batch_nb % 4) == 0)
```

File: pytorch_lightning/core/optimizer.py [hunk 11]
```diff
@@ -50,6 +49,14 @@ def __init__(self,
 
         self.__dict__ = {k: v for k, v in optimizer.__dict__.items() if k != 'step'}
 
+        # save functions and attributes from optimizer
+        self.state_dict = optimizer.state_dict
+        self.load_state_dict = optimizer.load_state_dict
+        self.zero_grad = optimizer.zero_grad
+        self.add_param_group = optimizer.add_param_group
+        self.__setstate__ = optimizer.__setstate__
+        self.__getstate__ = optimizer.__getstate__
```
Comment (awaelchli) [inline pytorch_lightning/core/optimizer.py:58]: Does that mean the line below `self.__class__ = type(...` doesn't work as we expect?
I thought it would handle that

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/optimizer.py (hunk 11 above)
Comment (ananyahjha93) [reply pytorch_lightning/core/optimizer.py:58]: LARS in bolts was failing when this wasn't included. LARS in bolts is just a wrapper around an optimizer, where the optimizer within LARSWrapper sets these params as well

Comment (tchaton) [reply pytorch_lightning/core/optimizer.py]: I was making some test. It works, just attributes where missing.

File: docs/source/optimizers.rst [hunk 12]
```diff
@@ -191,46 +191,69 @@ override the :meth:`optimizer_step` function.
...
+               optimizer.step(closure=closure)
 
         # update discriminator opt every 4 steps
         if optimizer_i == 1:
             if batch_nb % 4 == 0 :
-                optimizer.step()
-                optimizer.zero_grad()
+               optimizer.step(closure=closure)
+
+.. note:: When using ``Trainer(enable_pl_optimizer=True)``, ``.step`` accept a boolean ``make_optimizer_step`` which can be used as follow.
```
Comment (tchaton) [inline docs/source/optimizers.rst:213]:
```suggestion
.. note:: When using ``Trainer(enable_pl_optimizer=True)``, ``.step`` accepts a boolean ``make_optimizer_step`` which can be used as follow.
```
